src/basic_recipes/barplot.jl

```diff
@@ -46,15 +49,60 @@ function AbstractPlotting.plot!(p::BarPlot)
         end
     end
 
-    bars = lift(p[1], p.fillto, p.width, in_y_direction) do xy, fillto, width, in_y_direction
+    bars = lift(p[1], p.fillto, p.width, p.dodge, p.x_gap, p.dodge_gap, p.stack, in_y_direction) do xy, fillto, width, dodge, x_gap, dodge_gap, stack, in_y_direction
+      
+        #
+        if dodge === automatic
+            n_dodge = 1
+        else
+            n_dodge = length(unique(dodge))
+        end
+        
+        x = first.(xy)
+        y = last.(xy)
+        
         # compute half-width of bars
         if width === automatic
-            # times 0.8 for default gap
-            width = mean(diff(first.(xy))) * 0.8
-            width = ifelse(isfinite(width), width, 0.8)
+            x_unique = unique(filter(isfinite, x))
+            
+            if length(x_unique) == 1
+                width = 1.0
+            else
+                width = mean(diff(sort(x_unique)))
+            end
         end
+        
+        # --------------------------------
+        # ------------ Dodging -----------
+        # --------------------------------
+        dodge_width = scale_width(x_gap, dodge_gap, n_dodge)
+
+        if dodge === automatic
+            i_dodge = 1
+        else 
+            i_dodge = categoric_position.(dodge, Ref(categoric_labels(dodge)))
```

Comment:
This is the "incriminated line" for me. I think somehow it would be nice for the user to be able to pass `i_dodge` directly.